index.js

```diff
@@ -6,63 +6,54 @@ var glob = require('glob');
 var arrify = require('arrify');
 var pify = require('pify');
 
-function sortPatterns(patterns) {
-	patterns = arrify(patterns);
+var globP = pify(glob, Promise).bind(glob);
 
-	var positives = [];
-	var negatives = [];
+function isNegative(pattern) {
+	return pattern[0] === '!';
+}
 
-	patterns.forEach(function (pattern, index) {
-		var isNegative = pattern[0] === '!';
-		(isNegative ? negatives : positives).push({
-			index: index,
-			pattern: isNegative ? pattern.slice(1) : pattern
-		});
-	});
+function generateGlobTasks(patterns, opts) {
+	var globTasks = [];
 
-	return {
-		positives: positives,
-		negatives: negatives
-	};
-}
+	patterns = arrify(patterns);
+	opts = objectAssign({ignore: []}, opts);
 
-function setIgnore(opts, negatives, positiveIndex) {
-	opts = objectAssign({}, opts);
+	patterns.forEach(function (pattern, i) {
+		if (isNegative(pattern)) {
+			return;
+		}
 
-	var negativePatterns = negatives.filter(function (negative) {
-		return negative.index > positiveIndex;
-	}).map(function (negative) {
-		return negative.pattern;
+		var ignore = patterns.slice(i).filter(isNegative).map(function (pattern) {
+			return pattern.slice(1);
+		});
+
+		globTasks.push({
+			pattern: pattern,
+			opts: objectAssign({}, opts, {
+				ignore: opts.ignore.concat(ignore)
+			})
+		});
 	});
 
-	opts.ignore = (opts.ignore || []).concat(negativePatterns);
-	return opts;
+	return globTasks;
 }
 
 module.exports = function (patterns, opts) {
-	var sortedPatterns = sortPatterns(patterns);
-	opts = opts || {};
+	var globTasks = generateGlobTasks(patterns, opts);
 
-	if (sortedPatterns.positives.length === 0) {
```

Comment:
In the new `generateGlobTasks` logic, in this case it would map an empty array which is a no-op (returns another empty array), which is then passed to `Promise.all` and resolves to an empty array, right? Let's avoid unnecessary logic branches, It is better this way than having to add `Promise.resolve([])` in early returns. :smiley: